### ISSUE TITLE
pathSolver – return promise

### ISSUE THREAD

**Issue by Ralle** [flowtsohg/mdx-m3-viewer#50]
Hey cool library, I like it very much ;-).

I have a problem where models might refer to a texture with a .blp extension, but the way Warcraft 3 works, this path is equivalent with the same path with all these extensions ['dds', 'tif', 'tga', 'blp'] and I only know which one actually exists by the time the fetch is complete. So I would very much like to be able to have my pathSolver return a promise, potentially with just an extension and a buffer or a way to tell the viewer to auto detect the extension.

The only work-around I can think of to give me this is to preload a list of all paths before starting to load the model.

**Comment by flowtsohg** [comment]
I think extension detection is the only nice solution here, any thoughts of another solution? Didn't touch any code in a long time, but I suppose it's not much work to detect formats, especially for the binary ones.

**Comment by Ralle** [comment]
Nice to hear from you :).

Allowing the pathSolver to return a promise that resolves to the expected values the pathSolver usually gets would give me everything I need. But automatic is also cool.

**Comment by flowtsohg** [comment]
Thinking about this more, I am not sure how this could be implemented without changes to the entire loading, much like what I saw in the past.
It doesn't really matter which part returns a promise and which part waits, at the end of the day it means the actual resources are not available once the load function finishes.
This both means that all client code has to handle promises, which while disruptive isn't necessarily bad (it might be a more "correct" solution to loading), however it also means that all of the internal code needs to change, e.g. textures used by models are stored directly, doing this with promises will be a big mess.

**Comment by Ralle** [comment]
Okay. As I can already parse the model server-side, I think I may have to make a dictionary of URLs with their correct extension.

**Comment by Ralle** [comment]
Alternatively, you permit the extension to be a function that you execute on the response object / buffer / stream during the fetch.

**Comment by flowtsohg** [comment]
I can't actually think of any proper way to do this without forcing all loads to be promises.

By the time there is a response object, the load function already ended, probably seconds earlier.

Even if I wrap the implementation objects with something like:
```ts
class Resource {
    object: Model | Texture;
}
```
...the fact of the matter is, there is no implementation object when the load function returns, i.e. `object === null`, meaning this doesn't really give anything more than a roundabout implementation of promises.

How bad would you recon moving to promises will be?
```ts
let model = viewer.load(...);
let instance = model.addInstance();
// vs
viewer.load(...).then((model) => {
  let instance = model.addInstance();
});
// or (but of course this requires an async function which comes with its own side effects)
let model = await viewer.load(...);
let instance = model.addInstance();
```

It's probably more messy to do in the handlers with e.g. the mentioned texture arrays.

**Comment by flowtsohg** [comment]
One thing that is possible is to have a simple standardized Texture class that can support all of the different textures, and make texture loading automated.
It's not really possible to do the same with the models since they behave differently.

**Comment by Ralle** [comment]
That would be great too. As of right now the safest way to run the viewer is to have a complete index of all textures in the casc at hand because there’s no telling which one is dds, tga or blp. Then again I could make a list of everything that isn’t a dds. That would be shorter. But the point is that for texture paths they are all equivalent in-game so we have to be flexible too

**Comment by flowtsohg** [comment]
Perhaps something like this:
```ts
function pathSolver(src, solverParams) {
  return [`Resources/${src}`, true];
}

let promise = viewer.load('Units/Human/Footman/Footman.mdx', pathSolver);
promise.then((model) => {
  let instance = model.addInstance();
});
// promise of the first promise from a promise cache
// same as Promise.resolve(promise)
let promise2 = viewer.load('Units/Human/Footman/Footman.mdx', pathSolver);
// ...
// ...
// the model loaded already and is cached
// same as Promise.resolve(model)
let promise3 = viewer.load('Units/Human/Footman/Footman.mdx', pathSolver);
```
and that gives another side effect of no longer supporting events, but that's not necessarily bad.

**Comment by Ralle** [comment]
You are suggesting two things here, is that correct?

1. To auto detect texture formats.

and

2. To allow promises for loading models.

What I need is EITHER:

1. Auto detect models.

or

2. Allow promises for loading textures.

**Comment by flowtsohg** [comment]
I am proposing that all loads are promises, which adds the possibility of checking formats.
This means you could no longer do things like loading a texture and instantly adding it to an instance before it got loaded, or creating instances of models that didn't load, and so on.

**Comment by Ralle** [comment]
That sounds fine. But it is also fine if you do the minimal amount of work to solve my problem 😄

**Comment by flowtsohg** [comment]
There isn't really a way to do this without changing the entire loading code. It is something I thought about in the past and came to the same conclusion, but then there was less of a compelling reason to actually make the change.

**Comment by flowtsohg** [comment]
So what exactly are the rules when loading textures? Currently the handler replaces any extension with `.dds` when in reforged mode, is that actually needed?

I am still unsure how to handle text formats, which really just means MDL. If they are loaded with a fetch, the input is an ArrayBuffer, since there are no longer pre-known data types, but when loaded directly the input is meant to be a string.

Either way this change will take some time, because it cascades to tens of other files, and assumptions and conditions spread all over the place that need to be changed.

**Comment by flowtsohg** [comment]
Next issue - attachment models are a problem, because if their models didn't load before an instance of the parent model is created...well, they don't exist.
Having a concrete object with events that can be referenced is important when things load at different times, but there can't be a concrete type since the type is async.
It's somewhat amazing how such a simple thing you never even think about on desktop is such a huge issue when it goes async.
I think the only consistent solid solution that will actually support everything is both promises and a wrapper resource with events, something like:
```ts
let resource = viewer.load('Units/Human/Footman/Footman.mdx', pathSolver);

console.log(resource.h); // undefined

resource.whenLoaded()
  .then((model) => {
    console.log(model); // MdxModel { ... }
    console.log(model === resource.h) // true
  });

resource.on('load', (model) => {
  console.log(model); // MdxModel { ... }
  console.log(model === resource.h) // true
});
```
and that of course makes the usage messy.

I am not sure what a good solution is...

**Comment by flowtsohg** [comment]
I added an array of "blockers" to resources, where promises can be added when loading to stop the parent resource from finishing to load.
For now it's only used by attachments with model paths, since if a texture is missing, for example, a model can still render fine.
The changes are very yiffy so far, but things are becoming operational.

Still no clue what to do with MDL.

**Comment by Ralle** [comment]
Wouldn't this solve your problem?
`model.load("url", modelResolverAlsoForAttachmentsWhichReturnsPromise).then(...)`

You would call it with the model url and the included resolver would return a promise of the model. The loader then uses it for attachments and waits for those as well and recursively on their attachments, etc. Promise.all comes in handy here.

`TEXTURE_EXTENSIONS = ['tga', 'tif', 'blp', 'dds']`

`MODEL_EXTENSIONS = ['pkfx', 'pkb', 'mdl', 'mdx']`

Basically, whenever the system tries to load a file with an extension in the `TEXTURE_EXTENSIONS` array, it tries each of the extensions in the array in turn until it finds the texture. I guess it tries the original extension first, then from that order (I verified it with the devs).

For models it's the same thing, but the only extension I ever saw is mdx.

**Comment by flowtsohg** [comment]
I added the promise-based-loading branch. Not everything is working the same, nor am I happy with how some things look, but over all MDX should be fine for testing. Care to test it?

For now path solvers need to return `[finalSrc, isFetch]`, and of course `viewer.load(...)` returns a promise.

MDL testing is done by looking for the string `FormatVersion` in the first 4KB of either an ArrayBuffer or a string, depending how it gets loaded.
It's not ideal, but it seems to be quick enough to not really matter outside of my head.

I can't really test reforged/sc2 things because I don't have the files around.

**Comment by Ralle** [comment]
Yes! I will test it. But it might be a couple days. I hope to have time tomorrow but nothing is certain when you have offspring. Thanks a lot!

**Comment by flowtsohg** [comment]
`isFetch` is no longer needed, so path solvers just return whatever they want to load, i.e. `(src, solverParams) => finalSrc`.
I took advantage of the fact that testing for the format is really quick, so if the value is a string and doesn't match any handler, it is assumed to be a fetch.

The one thing that is still very yiffy is as I said at the start, the arrays of textures in models. I did a "sort of" design fix, but it's really ugly, need to think about it more.

**Comment by Ralle** [comment]
Okay I tried. Getting lots of errors. The model loads, but not textures. Here's the console:

<img width="830" alt="Screenshot 2020-11-08 at 14 57 47" src="https://user-images.githubusercontent.com/467309/98466938-ec863800-21d2-11eb-81a2-2b92a2f194ba.png">
<img width="841" alt="Screenshot 2020-11-08 at 14 57 58" src="https://user-images.githubusercontent.com/467309/98466939-edb76500-21d2-11eb-9d6e-c7af6b94beaa.png">

**Comment by flowtsohg** [comment]
I forgot to mention that handlers no longer implicitly depend on other handlers.
E.g. the MDX handler does not add the BLP/TGA handlers, you gotta do it manually.
This is to finally properly support the person that was using pre-converted textures to be used with wasm, which means he never even needed to support BLP or TGA.
So make sure to add them and the DDS handler.

Either way when logging errors, you probably want to print the 2nd and 3rd arguments, I am guessing if you did you would see the error saying it didn't find the handlers :P

**Comment by Ralle** [comment]
Oh great. I was waiting for an email notification about your reply but never got one, finally saw your message just now!

Cool that someone uses it in wasm.

**Comment by Ralle** [comment]
Worked after adding these:

```
			viewer.addHandler(handlers.mdx);
			viewer.addHandler(handlers.dds);
			viewer.addHandler(handlers.blp);
			viewer.addHandler(handlers.tga);
			viewer.addHandler(handlers.m3);
```

But m3 models aren't working. The promise returns null and emits an error.

I loaded this one:
https://www.hiveworkshop.com/threads/female-zealot.198041/#resource-66595

It never asks for assets, so you don't need any to reproduce it.

**Comment by flowtsohg** [comment]
M3 isn't ready, mostly still pondering what to do with textures for it and MDX. I also don't have any sc2 files roaming around.

**Comment by Ralle** [comment]
You can use this API:
https://www.hiveworkshop.com/mpq-contents?path=assets/textures/zealot_diffuse.dds

It allows cross origin.

**Comment by flowtsohg** [comment]
The API url is fine, but the urls it redirects to are not and do get blocked.

I believe M3 should work now as well, but still things to improve.

**Comment by flowtsohg** [comment]
I merged the new code with master. Everything should supposedly. If you encounter problems please open a new issue.
